### PR TITLE
improve collapseTemplates

### PR DESCRIPTION
**Pull request by lievenhey**
collapseTemplates no longer breaks with operator<< and operator<

### CODE REVIEW

**Comment by lievenhey** [comment]
Now it looks for the operator keyword if < or > is found.

File: src/util.cpp [hunk 1]
```diff
@@ -41,22 +41,55 @@
 #include "data.h"
 #include "settings.h"
 
-static QString collapseTemplate(const QString& str, int level)
+namespace {
+bool isOperator(const QStringRef& str, int pos)
```
[the remark this answers is not in the document]
Comment (lievenhey) [reply src/util.cpp]: thanks, I don't have much experience with text parsing

File: src/util.cpp [hunk 2]
```diff
@@ -41,24 +41,51 @@
...
+        if (c == QLatin1Char('>')) {
+            return true;
+        }
+
+        return false;
+    };
+
     QString output;
     output.reserve(str.size());
+    const auto operatorKeyword = QStringLiteral("operator");
```
Comment (milianw) [inline src/util.cpp:67]: QLatin1String

File: src/util.cpp [hunk 3]
```diff
@@ -41,24 +41,51 @@
...
             }
         } else if (c == QLatin1Char('>')) {
             depth--;
+        } else if (c == QLatin1Char('o') && str.midRef(i, operatorKeyword.size()) == operatorKeyword) {
+            i += operatorKeyword.size();
+            output.append(operatorKeyword);
+            for (; i < size && isSpaceOrAngelBracket(str[i]); i++) {
+                output.append(str[i]);
+            }
+            output.append(str[i]);
```
Comment (milianw) [inline src/util.cpp:85]: why is this line here, this seems to be eating one char too many? are you trying to compensate for the `i++` in the inner and outer `for` loops? maybe just do a `--i` after the last iteration of the inner loop? or use different loop vars?

File: src/util.cpp [hunk 4]
```diff
@@ -41,24 +41,51 @@
...
 
-static QString collapseTemplate(const QString& str, int level)
+QString collapseTemplate(const QString& str, int level)
 {
+
     if (str.indexOf(QLatin1Char('<')) == -1) {
         return str;
     }
 
+    auto isSpaceOrAngelBracket = [](QChar c) {
```
Comment (milianw) [inline src/util.cpp:51]: Angle not :angel:

**Review by milianw — CHANGES_REQUESTED** [review]
nearly done ;-)

File: src/util.cpp (hunk 3 above)
Comment (lievenhey) [reply src/util.cpp:85]: I used a different loop var and added a comment explaining whats going on

**Comment by lievenhey** [comment]
should be done now

**Comment by milianw** [comment]
thanks!
